Orange/data/table.py

```diff
@@ -33,146 +28,176 @@ def get_sample_datasets_dir():
+# noinspection PyPep8Naming
+class Table(Storage, DataFrame):
+    _WEIGHTS_COLUMN = ContinuousVariable("__weights__")
+    _WEIGHTS_COLUMN.is_weight = True
 
-    def __repr__(self):
-        return self._str(True)
+    # a counter for indexing rows, important for deterministically selecting rows
+    # and keeping pandas indices sane
+    _next_instance_id = 0
+    _next_instance_lock = Lock()
 
+    __file__ = None
 
-class Columns:
-    def __init__(self, domain):
-        for v in chain(domain, domain.metas):
-            setattr(self, v.name.replace(" ", "_"), v)
+    # custom properties, preserved through pandas manipulations
+    _metadata = ['name', '__file__', '_columns_X', '_columns_Y', '_columns_meta']
 
+    @property
+    def _constructor(self):
+        """Proper pandas extension as per http://pandas.pydata.org/pandas-docs/stable/internals.html"""
+        return Table
 
-# noinspection PyPep8Naming
-class Table(MutableSequence, Storage):
-    __file__ = None
-    name = "untitled"
+    @property
+    def _constructor_sliced(self):
+        """Proper pandas extension as per http://pandas.pydata.org/pandas-docs/stable/internals.html"""
+        return TableSeries
+
+    @property
+    def _constructor_expanddim(self):
+        return TablePanel
+
+    @property
+    def domain(self):
+        # TODO: probably return a Domain object constructed from columns
+        return self.columns
+
+    @property
+    def columns_X(self):
+        """A read-only list of X column variables."""
+        return [c for c in self.columns if c in self._columns_X]
 
     @property
-    def columns(self):
+    def columns_Y(self):
+        """A read-only list of Y column variables."""
+        return [c for c in self.columns if c in self._columns_Y]
+
+    @property
+    def columns_meta(self):
+        """A read-only list of meta column variables."""
+        return [c for c in self.columns if c in self._columns_meta]
+
+    def _to_numpy(self, X=False, Y=False, meta=False, writable=False):
         """
-        A class whose attributes contain attribute descriptors for columns.
-        For a table `table`, setting `c = table.columns` will allow accessing
-        the table's variables with, for instance `c.gender`, `c.age` ets.
-        Spaces are replaced with underscores.
+        Exports a numpy matrix. The order is always X, Y, meta.
+        The columns are in the same order as in
+        Table.columns, Table.columns_X, Table.columns_Y, Table.columns_meta
+        If writable == False (default), the numpy writable flag is set to false.
+            This means write operations on this array will loudly fail. 
         """
-        return Columns(self.domain)
+        # TODO: only return numeric values here, need to transform
+        cols = []
+        cols += self.columns_X if X else []
+        cols += self.columns_Y if Y else []
+        cols += self.columns_meta if meta else []
+        res = self[cols].values
```

Comment:
Ah, yes, sure, categoricals need to be transformed _here in this function_. `table.X` is traditionally a numpy array of floats.